### ISSUE TITLE
incompatible_remove_exec_tools

### ISSUE THREAD

**Issue by katre** [bazelbuild/bazel#19132]
As of the Bazel 6.X series, there has been no semantic difference between `genrule`'s `tools` and `exec_tools` attributes. This incompatible change is to remove the `exec_tools` attribute and leave `tools` as the canonical attribute.

The fix for this is very simple: any `genrule` that uses the `exec_tools` attribute should instead use `tools`.

**Comment by katre** [comment]
Using this to track the fixes uncovered by #19130.

**Comment by meteorcloudy** [comment]
@sgowroji Can you help file issues against broken downstream projects and link back to this one?

**Comment by katre** [comment]
As a note, the failure in Bazel's own tests appears to be due to https://github.com/google/desugar_jdk_libs/blob/master/BUILD

**Comment by katre** [comment]
According to https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/1600, this flag is safe to flip and release. This is a problem, since the thing that actually needs to happen is to remove the `exec_tools` attribute entirely, and we know (from #19130) that this causes massive failures.

I'm going to revert the flag (but leave `exec_tools` in Bazel's `genrule`), and keep this issue to track the process of removing uses of the attribute.

**Comment by katre** [comment]
Okay, I ran downstream tests based on completely removing the attribute (https://github.com/bazelbuild/bazel/pull/19142): https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/3207#_

Of these, there are only four actual failing projects due to this change:
- Bazel
  - actually desugar_jdk_libs
    - https://github.com/google/desugar_jdk_libs/pull/8
- Envoy
  - Already fixed
    - https://github.com/envoyproxy/envoy/issues/28719
- Kythe
  - https://github.com/kythe/kythe/issues/5771
- rules_haskell
  - https://github.com/tweag/rules_haskell/issues/1924

**Comment by meteorcloudy** [comment]
Perfect, thanks!

**Comment by katre** [comment]
This is fixed (as long as 5ce12ee isn't rolled back).

**Comment by illicitonion** [comment]
@katre Is there a reason the incompatible flag was not working or couldn't be introduced?

Given folks are using e.g. many versions of protobuf out in the wild, I'm pretty concerned that requiring people to be using protobuf >= 24.0 to use protobuf with bazel is going to cause some problems for Bazel 7 adoption. Having an incompatible flag for 7.x and removing in 8.x feels like smoother path?

**Comment by katre** [comment]
The problem is that bazel creates native rules a single time, at server startup, so there's no way to have a non-startup flag control whether an attribute exists. It would have been possible to use a startup flag, but that would cause a lot of server flapping, and be very difficult to plumb into the right places.

I already held off of removing `exec_tools` in the 6.X series: adding yet another Bazel LTS version with this misfeature is a non-starter. Anyone who cannot upgrade to a version of protobuf with this very simple change can continue to use the 6.X series until they can upgrade.

**Comment by illicitonion** [comment]
What was wrong with the flag you already introduced which, while not technically removing the attribute, had the effect of doing so as far as users are concerned, by hard erroring if it was set?

Given how small and not-affecting-anything-else the actual removal is in 5ce12ee3363c62218f096d16bd2424a86ebdae19, it feels like leaving those ~dozen isolated lines of code in Bazel for a while longer isn't too much of a hardship maintenance-wise?

Realistically, my large company is going to move to Bazel 7 as a whole, so if we have to delay the whole company from migrating to Bazel 7 for several months because one team has a difficult-to-avoid arcane protobuf dependency, that's going to be a fairly frustrating situation...

**Comment by katre** [comment]
The flag didn't work: there were genrules that set `exec_tools` and didn't get an error.

I understand the frustration you're feeling, but (as I said) this has already been delayed a year, and I spent about half that year cleaning up uses of `exec_tools` in google's internal codebase. Commit 5ce12ee3363c62218f096d16bd2424a86ebdae19 was small because of that effort. Reverting the changes runs the risk of projects (inside google and outside) starting to use `exec_tools` again, which just resets the clock.

If there's a bug with protobuf specifically I will be happy to look at it, if the problem is that users can't migrate easily to a version of protobuf that removes `exec_tools` that's unfortunate. One approach I had to take inside Bazel was 6a313f157d440877e96cca83b4e95532dc09527f, which added a small patch to replace `exec_tools` with `tools` in the offending BUILD files. Perhaps that will help your protobuf cleanup?
